recitation_6/file-io.md

```diff
@@ -0,0 +1,234 @@
+-   **(2) stderr (standard error)**  This stream is for error messages and is not buffered, meaning any characters written to it will immediately be flushed to their destination. This destination is often the terminal screen, but can be other locations as well.
+
+If you wish to interact with these buffers you will need to  `#include <stdio.h>`. This library defines standard functions such as  `printf()`  `scanf()`  and others which you may or may not have already used.
+
+### Redirection and Piping
+
+Lots of input and output to/from programs is visible in the shell (your terminal screen). That being said, it's possible to redirect the source of stdin or the destination of stderr and stdout.
+
+There are several kinds of redirection possible from the console:
+
+-   `>`  redirects the standard output of the left argument to the right argument. For example  `echo "hello world" > hi.txt`  will write the words 'hello world' to a file called  `hi.txt`, and 'hello world' will not appear on the terminal screen. 
+-   `2>`  is similar, except that it takes the standard error and redirects it, rather than the standard output.
+-   `<`  does 'input redirection', meaning that the contents of the file on the right of the operator serves as the standard input of the left argument. For example, after running the example above,  `cat < hi.txt`  will print "hello world" to the console.
+-   `2>&1`  will redirect stderr to the same location as stdout.
+-   `>>`  will append the output to a file instead of overwriting the file.
+-  `|`  will feed the output from the program on the left as input to the program on the right. For example the `head -n` command will print as output the first n number of data of its input. Running `ls | head -3` will print up to 3 files in your working directory. 
+
+### Formatting
+
+printf and scanf both use format strings to specify what how to format their output. They also both accept variable arguments. All arguments to scanf  **must**  be pointers whereas arguments to printf should be values (in the case of numbers) or  `char *`  in the case of strings. Pages 153-154 in the K&R explain how to format your format strings for  `printf()`  and 157-158 explain formatting for  `scanf()`. Make sure you can identify the following two format strings:
+
+	printf("%-15.10s", "hello, world");
+	sscanf("25 Dec 1988", "%d %s %d", &day, month, &year);
+
+
+### FILE Pointers
+
+`FILE`  is a typedef'ed structure in  `stdio.h`. Whenever you use it, you'll use a  `FILE *`  though because you'll always be getting a value back from/passing it to common file operating functions. Why a pointer? Because these functions will modify the internal values of the  `FILE`  value. So while you could copy them because it's a struct and C would be fine with passing it by value, things like your place in the file would not be maintained. A  `FILE *`  is given as an  _opaque type_, meaning you should never be accessing it's internals directly.
```

Comment:
I think it would be much better to get rid of all of this:
"Why a pointer? Because these functions will modify the internal values of the `FILE` value. So while you could copy them because it's a struct and C would be fine with passing it by value, things like your place in the file would not be maintained."